service/src/main/java/bio/terra/cbas/runsets/monitoring/SmartRunsPoller.java

```diff
@@ -0,0 +1,78 @@
+package bio.terra.cbas.runsets.monitoring;
+
+import static java.util.stream.Collectors.groupingBy;
+
+import bio.terra.cbas.dao.RunDao;
+import bio.terra.cbas.dependencies.wes.CromwellService;
+import bio.terra.cbas.model.RunState;
+import bio.terra.cbas.models.Run;
+import cromwell.client.ApiException;
+import cromwell.client.model.State;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+@Component
+public class SmartRunsPoller {
+
+  private final CromwellService cromwellService;
+  private final RunDao runDao;
+
+  private static final org.slf4j.Logger logger = LoggerFactory.getLogger(SmartRunsPoller.class);
+
+  public SmartRunsPoller(CromwellService cromwellService, RunDao runDao) {
+    this.cromwellService = cromwellService;
+    this.runDao = runDao;
+  }
+
+  public List<Run> updateRuns(List<Run> runs) {
+    // Filter only updatable runs:
+    Stream<Run> updatableRuns = runs.stream().filter(this::nonTerminal);
+
+    // Group by current (engine) status:
+    Map<State, List<Run>> engineStatus =
+        updatableRuns.collect(
+            groupingBy(
+                r -> {
+                  try {
+                    return cromwellService.runStatus(r.engineId()).getState();
```

Comment:
If there are a lot of statuses to check this could possibly be improved by calling `runStatus()` in parallel. replacing `groupingBy()` with `groupingByConcurrent()` should do this.

Reply:
For context - it's known and accepted that this v0.2 is not going to scale well... I agree that polling in parallel is probably in our future, but perhaps by batching bigger questions to the engine rather than asking all the individual questions in parallel. In other words... tbd!

(Very glad to find out about `groupingByConcurrent()` though!)